File: lab_lectures/lab4/manual_mem.py
import gc
import weakref
import time

# Disable garbage collector
gc.disable()

class MemoryHungryObject:
    def __init__(self, name, data):
        self.name = name
        # Create a large list to consume memory
        self.data = [data] * 1000000

def create_objects():
    # Dictionary to store our objects
    objects = {}
    
    # Create several large objects
    for i in range(5):
        objects[f"object_{i}"] = MemoryHungryObject(f"object_{i}", i)
        print(f"Created {objects[f'object_{i}'].name}")
    
    return objects

def main():
    print("Garbage collector enabled:", gc.isenabled())
    print("Creating objects...")
    
    # Create objects and store them
    objects = create_objects()
    
    # Let's see the objects we've created
    print("\nBefore manual cleanup:")
    print("Current objects:", list(objects.keys()))
    
    # Manual cleanup of specific objects
    print("\nPerforming manual cleanup...")
    del objects["object_1"]
    del objects["object_3"]
    
    # Force garbage collection despite it being disabled
    gc.collect()
    
    print("\nAfter manual cleanup:")
    print("Remaining objects:", list(objects.keys()))
    
    # Remove the rest of the objects
    del objects["object_0"]
    del objects["object_2"]
    del objects["object_4"]
    
    # Force garbage collection despite it being disabled
    gc.collect()
    
    print("\nAfter manual cleanup two:")
    print("Remaining objects:", list(objects.keys()))
    
    # Re-enable garbage collector before program ends
    gc.enable()
    print("\nGarbage collector re-enabled:", gc.isenabled())

if __name__ == "__main__":
    main()

File: lab_lectures/lab4/manual_mem.cpp
#include <iostream>
#include <string>
#include <vector>

#define OBJECT_COUNT 5
#define DATA_SIZE 1000000

class MemoryHungryObject {
public:
    std::string name;
    int *data;

    MemoryHungryObject(const std::string &name, int value) : name(name) {
        data = new int[DATA_SIZE];
        for (int i = 0; i < DATA_SIZE; i++) {
            data[i] = value;
        }
    }

    ~MemoryHungryObject() {
        delete[] data;
    }
};

void create_objects(std::vector<MemoryHungryObject*> &objects) {
    for (int i = 0; i < OBJECT_COUNT; i++) {
        std::string name = "object_" + std::to_string(i);
        objects.push_back(new MemoryHungryObject(name, i));
        std::cout << "Created " << objects.back()->name << std::endl;
    }
}

void print_memory_usage(const std::vector<MemoryHungryObject*> &objects) {
    std::cout << "\nObjects memory usage\n";
    for (const auto &obj : objects) {
        if (obj != nullptr) {
            std::cout << obj->name << ": " << sizeof(int) * DATA_SIZE << " Bytes" << std::endl;
        }
    }
}

int main() {
    std::vector<MemoryHungryObject*> objects;

    std::cout << "Creating objects...\n";
    create_objects(objects);
    print_memory_usage(objects);

    std::cout << "\nBefore manual cleanup:\n";
    std::cout << "Current objects: ";
    for (const auto &obj : objects) {
        if (obj != nullptr) {
            std::cout << obj->name << " ";
        }
    }
    std::cout << "\n";

    std::cout << "\nPerforming manual cleanup...\n";
    delete objects[1];
    objects[1] = nullptr;
    delete objects[3];
    objects[3] = nullptr;

    std::cout << "\nRemaining objects memory usage:\n";
    for (const auto &obj : objects) {
        if (obj != nullptr) {
            std::cout << obj->name << ": " << sizeof(int) * DATA_SIZE << " Bytes" << std::endl;
        }
    }

    std::cout << "\nAfter manual cleanup:\n";
    std::cout << "Remaining objects: ";
    for (const auto &obj : objects) {
        if (obj != nullptr) {
            std::cout << obj->name << " ";
        }
    }
    std::cout << "\n";

    for (auto &obj : objects) {
        if (obj != nullptr) {
            delete obj;
            obj = nullptr;
        }
    }

    std::cout << "\nAfter manual cleanup two:\n";
    std::cout << "Remaining objects: None\n";

    return 0;
}